drivers/stepper/ti/drv8424.c

```diff
@@ -0,0 +1,680 @@
+
+/**
+ * @brief DRV8424 stepper driver configuration data.
+ *
+ * This structure contains all of the devicetree specifications for the pins
+ * needed by a given DRV8424 stepper driver.
+ */
+struct drv8424_config {
+	struct gpio_dt_spec dir_pin;
+	struct gpio_dt_spec step_pin;
+	struct gpio_dt_spec sleep_pin;
+	struct gpio_dt_spec en_pin;
+	struct gpio_dt_spec m0_pin;
+	struct gpio_dt_spec m1_pin;
+	const struct device *counter;
+};
+
+/* Struct for storing the states of output pins. */
+struct drv8424_pin_states {
+	uint8_t sleep: 1;
+	uint8_t en: 1;
+	uint8_t m0: 2;
+	uint8_t m1: 2;
+};
+
+/**
+ * @brief DRV8424 stepper driver data.
+ *
+ * This structure contains mutable data used by a DRV8424 stepper driver.
+ */
+struct drv8424_data {
+	const struct device *dev;
+	bool enabled;
+	struct drv8424_pin_states pin_states;
+	enum stepper_micro_step_resolution ustep_res;
+	uint32_t max_velocity;
+	int32_t reference_position;
+	int32_t target_position;
+	bool is_moving;
+	bool constant_velocity;
+	enum stepper_direction constant_velocity_direction;
+	stepper_event_callback_t event_callback;
+	struct k_work event_callback_work;
+	void *event_callback_user_data;
+	struct k_msgq event_msgq;
+	char event_msgq_buffer[CONFIG_DRV8424_EVENT_QUEUE_LEN * sizeof(enum stepper_event)];
+	bool step_signal_high;
+	struct counter_top_cfg counter_top_cfg;
+};
+
+static void drv8424_user_callback_work_fn(struct k_work *work)
+{
+	int ret;
+
+	/* Get containing data struct */
+	struct drv8424_data *driver_data =
+		CONTAINER_OF(work, struct drv8424_data, event_callback_work);
+
+	/* Get next pending event (if any) from message queue */
+	enum stepper_event event;
+
+	ret = k_msgq_get(&driver_data->event_msgq, &event, K_NO_WAIT);
+	if (ret != 0) {
+		/* No pending events */
+		return;
+	}
+
+	/* Run the callback */
+	if (driver_data->event_callback != NULL) {
+		driver_data->event_callback(driver_data->dev, event,
+					    driver_data->event_callback_user_data);
+	}
+
+	/* If there are more pending events, resubmit this work item to handle them */
+	if (k_msgq_num_used_get(&driver_data->event_msgq) > 0) {
+		k_work_submit(work);
+	}
+}
+
+static int drv8424_schedule_user_callback(struct drv8424_data *data, enum stepper_event event)
+{
+	int ret;
+
+	/* Place event in message queue to be picked up by work item */
+	ret = k_msgq_put(&data->event_msgq, &event, K_NO_WAIT);
+	if (ret != 0) {
+		LOG_WRN("%s: Too many concurrent events, dropping event of type %d",
+			data->dev->name, event);
+		return -ENOBUFS;
+	}
+
+	/* Submit work item */
+	ret = k_work_submit(&data->event_callback_work);
+	if (ret < 0) {
+		LOG_ERR("%s: Failed to submit work item (error %d)", data->dev->name, ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static void drv8424_positioning_top_interrupt(const struct device *dev, void *user_data)
+{
+	struct drv8424_data *data = (struct drv8424_data *)user_data;
+	const struct drv8424_config *config = (struct drv8424_config *)data->dev->config;
```

Comment:
no need for these castings, both are `void *` it's not like the compiler can do anything about it anyway, drop